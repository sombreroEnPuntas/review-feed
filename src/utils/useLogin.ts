import { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import {
  clearToken,
  getError,
  getToken,
  initialToken,
  setToken,
} from '../data/login'

const useLogin = () => {
  const dispatch = useDispatch()

  const error: string = useSelector(getError)
  const token: string = useSelector(getToken)

  useEffect(() => {
    dispatch(initialToken())
  }, [dispatch])

  const login = (username: string, password: string) => {
    dispatch(setToken(username, password))
  }

  const logout = () => {
    dispatch(clearToken())
  }

  return { error, login, logout, token }
}

export default useLogin
