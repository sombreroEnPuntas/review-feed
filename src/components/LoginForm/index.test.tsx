import { render, fireEvent } from '@testing-library/react'
import React from 'react'
import { ThemeWrapper } from 'retro-ui'

import TestedComponent from '.'

// Deps
import useLogin from '../../utils/useLogin'

// Mocks
jest.mock('../../utils/useLogin')

// Mock data
const setUseLoginMock = ({
  error,
  loading,
  login = jest.fn(),
  logout = jest.fn(),
  token,
}) =>
  (useLogin as jest.Mock).mockImplementation(() => ({
    error,
    loading,
    login,
    logout,
    token,
  }))

describe('LoginForm', () => {
  afterEach(() => {
    // much force-casting, such type-safe, so dark side
    ;(useLogin as jest.Mock).mockClear()
  })
  afterAll(() => {
    jest.resetAllMocks() // clean .mock
  })

  it(`renders a loading view`, async () => {
    const testData = {
      error: null,
      loading: true,
      token: undefined,
    }
    setUseLoginMock(testData)

    const { container, getByText } = render(
      <ThemeWrapper>
        <TestedComponent />
      </ThemeWrapper>
    )
    const loadingText = getByText(/Loading/)

    expect(loadingText).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it(`renders a logged-in view`, async () => {
    const testData = {
      error: null,
      loading: false,
      token: 'ðŸ¦•',
    }
    setUseLoginMock(testData)

    const { container, getByText } = render(
      <ThemeWrapper>
        <TestedComponent />
      </ThemeWrapper>
    )
    const logoutButton = getByText(/Logout/)

    expect(logoutButton).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it(`renders a login form`, async () => {
    const testData = {
      error: null,
      loading: false,
      token: undefined,
    }
    setUseLoginMock(testData)

    const { container, getByText } = render(
      <ThemeWrapper>
        <TestedComponent />
      </ThemeWrapper>
    )
    const loginButton = getByText(/Login/)

    expect(loginButton).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it(`renders a login form with error`, async () => {
    const testData = {
      error: `418: The resulting entity body MAY be short and stout.`,
      loading: false,
      token: undefined,
    }
    setUseLoginMock(testData)

    const { container, getByText } = render(
      <ThemeWrapper>
        <TestedComponent />
      </ThemeWrapper>
    )
    const loginButton = getByText(/Login/)

    expect(loginButton).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it(`reacts to login cta`, async () => {
    const login = jest.fn()
    const testData = {
      error: null,
      loading: false,
      login,
      token: undefined,
    }
    setUseLoginMock(testData)

    const { getByText, getByLabelText } = render(
      <ThemeWrapper>
        <TestedComponent />
      </ThemeWrapper>
    )
    const loginButton = getByText(/Login/)
    const usernameInput = getByLabelText(/username/)
    const passwordInput = getByLabelText(/password/)

    fireEvent.change(usernameInput, { target: { value: 'Sally' } })
    fireEvent.change(passwordInput, { target: { value: 'Rawr!' } })
    fireEvent.click(loginButton)

    expect(login).toBeCalledWith('Sally', 'Rawr!')
  })

  it(`reacts to logout cta`, async () => {
    const logout = jest.fn()
    const testData = {
      error: null,
      loading: false,
      logout,
      token: 'ðŸ¦•',
    }
    setUseLoginMock(testData)

    const { getByText } = render(
      <ThemeWrapper>
        <TestedComponent />
      </ThemeWrapper>
    )
    const logoutButton = getByText(/Logout/)

    fireEvent.click(logoutButton)

    expect(logout).toBeCalledWith()
  })
})
