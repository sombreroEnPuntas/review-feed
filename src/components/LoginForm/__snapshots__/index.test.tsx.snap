// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders a logged-in view 1`] = `
.c0 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
  background-color: #E4303E;
  box-shadow: none;
}

<div>
  <button
    class="sc-AxiKw c0"
  >
    Logout
  </button>
</div>
`;

exports[`LoginForm renders a login form 1`] = `
.c2 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  display: block;
  margin: 0.5em;
  max-width: 450px;
  overflow-x: auto;
  padding: 0.5em;
  white-space: pre-wrap;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #212529;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  line-height: calc(2 * 1.5em);
}

.c8 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
}

.c5 {
  background-color: white;
  border: none;
  box-shadow: 0 4px #4d5256, 0 -4px #4d5256, 4px 0 #4d5256, -4px 0 #4d5256;
  font-family: 'Press Start 2P',monospace;
  line-height: calc(2 * 1.5em);
  margin: 4px;
  outline-color: #212529;
  padding: 0.5rem 1rem;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #4d525666;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  left: 0;
  overflow: auto;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c7 {
  font-size: 1em;
  line-height: 1.5em;
  margin: 1em 0;
  min-height: 3em;
}

.c6 {
  margin: 1em 0 0 0;
}

.c0 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
  background-color: #E4303E;
  box-shadow: none;
}

.c4 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

<div>
  <button
    class="c0"
  >
    Logout
  </button>
  <div
    class="c1"
  >
    <pre
      class="c2"
      type="black"
    >
      <code
        class="c3"
        type="black"
      >
        <form
          class="c4"
        >
          <input
            aria-label="username"
            autocomplete="username"
            class="c5"
            name="username"
            placeholder="username"
            required=""
            type="text"
            value=""
          />
          <div
            class="c6"
          />
          <input
            aria-label="password"
            autocomplete="password"
            class="c5"
            name="password"
            placeholder="password"
            required=""
            type="password"
            value=""
          />
          <div
            class="c7"
          >
            


          </div>
          <button
            class="c8"
          >
            Login
          </button>
        </form>
      </code>
    </pre>
  </div>
</div>
`;

exports[`LoginForm renders a login form with error 1`] = `
.c2 {
  background-color: #ffd7cf;
  box-shadow: 4px 0 #e76e55, 0 -4px #e76e55, -4px 0 #e76e55, 0 4px #e76e55;
  display: block;
  margin: 0.5em;
  max-width: 450px;
  overflow-x: auto;
  padding: 0.5em;
  white-space: pre-wrap;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #ce372b;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  line-height: calc(2 * 1.5em);
}

.c8 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
}

.c5 {
  background-color: white;
  border: none;
  box-shadow: 0 4px #4d5256, 0 -4px #4d5256, 4px 0 #4d5256, -4px 0 #4d5256;
  font-family: 'Press Start 2P',monospace;
  line-height: calc(2 * 1.5em);
  margin: 4px;
  outline-color: #212529;
  padding: 0.5rem 1rem;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #4d525666;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  left: 0;
  overflow: auto;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1;
}

.c7 {
  background-color: #ffd7cf;
  color: #ce372b;
  font-size: 1em;
  line-height: 1.5em;
  margin: 1em 0;
  min-height: 3em;
}

.c6 {
  margin: 1em 0 0 0;
}

.c0 {
  background-color: white;
  box-shadow: 4px 0 #4d5256, 0 -4px #4d5256, -4px 0 #4d5256, 0 4px #4d5256;
  color: #212529;
  cursor: pointer;
  display: block;
  font-family: 'Press Start 2P',monospace;
  font-size: 0.65em;
  font-weight: bold;
  line-height: calc(2 * 1.5em);
  margin: 0.5em;
  padding: 0.5em;
  background-color: #E4303E;
  box-shadow: none;
}

.c4 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

<div>
  <button
    class="c0"
  >
    Logout
  </button>
  <div
    class="c1"
  >
    <pre
      class="c2"
      type="error"
    >
      <code
        class="c3"
        type="error"
      >
        <form
          class="c4"
        >
          <input
            aria-label="username"
            autocomplete="username"
            class="c5"
            name="username"
            placeholder="username"
            required=""
            type="text"
            value=""
          />
          <div
            class="c6"
          />
          <input
            aria-label="password"
            autocomplete="password"
            class="c5"
            name="password"
            placeholder="password"
            required=""
            type="password"
            value=""
          />
          <div
            class="c7"
          >
            418: The resulting entity body MAY be short and stout.
          </div>
          <button
            class="c8"
          >
            Login
          </button>
        </form>
      </code>
    </pre>
  </div>
</div>
`;
